### PR TITLE
Add ability to track additional info in tokens

### PR DESCRIPTION
**Pull request by mhfs**
Hey Adam!

First of all, thanks for all your work on Tiddle.

In our product, we have the need of tracking some additional context information when issuing authentication tokens.

This PR adds the ability to pass a `metadata` hash to `create_and_return_token` so that arbitrary data can be stored in the token.

I don't expect it to cause any backward compatibility issues.

Looking forward to your thoughts!

### CODE REVIEW

**Comment by mhfs** [comment]
@adamniedzielski it seems appraisal has an unresolved issue with ruby 3. I believe that's why some checks failed.

https://github.com/thoughtbot/appraisal/issues/186

**Comment by mhfs** [comment]
@adamniedzielski anything else I can do to help this move along?

**Comment by adamniedzielski** [comment]
Hey @mhfs 👋  thank you for your contribution!

Do you think that you could satisfy your usecase in the following way:

```ruby
token = Tiddle.create_and_return_token(user, request)
token.update!(custom_field: custom_value)
```

? This keeps the code of the library simpler 😉

**Comment by mhfs** [comment]
hey @adamniedzielski, thanks for the reply and I totally understand the desire to keep the library small.

However, I explored your suggestion, and `Tiddle.create_and_return_token` returns a string and not the ActiveRecord object.

Also, by splitting the process into two steps, I'd lose the ability to do model validations on my custom attributes as they'd fail on the `create!` call Tiddle makes internally.

Any other ideas? :D

**Comment by mhfs** [comment]
hey @adamniedzielski, wonder if you had the chance of looking into my previous comment here. :)

**Comment by adamniedzielski** [comment]
> Tiddle.create_and_return_token returns a string and not the ActiveRecord object.

That's true, I totally overlooked this, sorry! And changing this behaviour would be a backwards incompatible change for all of the applications 😢 

> Also, by splitting the process into two steps, I'd lose the ability to do model validations on my custom attributes as they'd fail on the create! call Tiddle makes internally.

That's also a good point 👍 

I'm convinced to accept the PR after your arguments 😄

File: lib/tiddle/token_issuer.rb [hunk 1]
```diff
@@ -56,13 +56,13 @@ def authentication_token_class(resource)
       end
     end
 
-    def token_attributes(token_body, request, expires_in)
+    def token_attributes(token_body, request, expires_in, metadata: {})
```
Comment (adamniedzielski) [inline lib/tiddle/token_issuer.rb:59]: I think that it's time to turn this method to accept only keyword arguments, as we're adding the fourth argument.

File: spec/tiddle_spec.rb [hunk 2]
```diff
@@ -38,6 +38,11 @@
                                      FakeRequest.new(user_agent: "Internet Explorer 4.0")
       expect(@user.authentication_tokens.last.user_agent).to eq "Internet Explorer 4.0"
     end
+
+    it "saves additional metadata" do
+      Tiddle.create_and_return_token @user, FakeRequest.new, metadata: { metadata_attr1: "abc" }
+      expect(@user.authentication_tokens.last.metadata_attr1).to eq "abc"
+    end
```
Comment (adamniedzielski) [inline spec/tiddle_spec.rb:45]: Thank you for adding a test 👍

**Comment by adamniedzielski** [comment]
I tested the build problems in https://github.com/adamniedzielski/tiddle/pull/77 and they seem to be resolved by now 😄

File: lib/tiddle/token_issuer.rb (hunk 1 above)
Comment (mhfs) [reply lib/tiddle/token_issuer.rb]: :+1: done!

**Comment by mhfs** [comment]
awesome, @adamniedzielski. thanks for your comments and consideration, we really appreciate it. 

I made the keywords args change you suggested.

From my perspective, this is good to go. Anything else I can help with before the merge?

**Comment by adamniedzielski** [comment]
Awesome, thank you for your contribution! 🙏🤗 I’ll release a new version later.

**Pull request merged**